test/taskrun_test.go

```diff
@@ -209,3 +225,218 @@ spec:
+		t.Errorf("Error waiting for TaskRun to start running: %s", err)
+	}
+
+	patches := []jsonpatch.JsonPatchOperation{{
+		Operation: "replace",
+		Path:      "/status/taskSpec/steps/0/image",
+		Value:     "not-ubuntu",
+	}}
+	patchBytes, err := json.Marshal(patches)
+	if err != nil {
+		t.Fatalf("failed to marshal patch bytes in order to stop")
+	}
+	t.Logf("Patching TaskRun %s in namespace %s mid run for spire to catch the un-authorized changed", taskRunName, namespace)
+	if _, err := c.V1beta1TaskRunClient.Patch(ctx, taskRunName, types.JSONPatchType, patchBytes, metav1.PatchOptions{}, "status"); err != nil {
+		t.Fatalf("Failed to patch taskrun `%s`: %s", taskRunName, err)
+	}
+
+	t.Logf("Waiting for TaskRun %s in namespace %s to succeed", taskRunName, namespace)
+	if err := WaitForTaskRunState(ctx, c, taskRunName, TaskRunFailed(taskRunName), "TaskRunFailed", v1beta1Version); err != nil {
+		t.Errorf("Error waiting for TaskRun to finish: %s", err)
+	}
+
+	taskrun, err := c.V1beta1TaskRunClient.Get(ctx, taskRunName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Couldn't get expected TaskRun %s: %s", taskRunName, err)
+	}
+
+	if isSpireEnabled(ctx, t, c, namespace) {
+		spireShouldFailTaskRunResultsVerify(t, taskrun)
+		spireShouldFailSpireAnnotation(t, taskrun)
+	}
+
+	expectedStepState := []v1beta1.StepState{{
+		ContainerState: corev1.ContainerState{
+			Terminated: &corev1.ContainerStateTerminated{
+				ExitCode: 1,
+				Reason:   "Error",
+			},
+		},
+		Name:          "unnamed-0",
+		ContainerName: "step-unnamed-0",
+	}}
+
+	ignoreTerminatedFields := cmpopts.IgnoreFields(corev1.ContainerStateTerminated{}, "StartedAt", "FinishedAt", "ContainerID")
+	ignoreStepFields := cmpopts.IgnoreFields(v1beta1.StepState{}, "ImageID")
+	if d := cmp.Diff(taskrun.Status.Steps, expectedStepState, ignoreTerminatedFields, ignoreStepFields); d != "" {
+		t.Fatalf("-got, +want: %v", d)
+	}
+}
+
+func TestTaskRunSpire(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var c *clients
+	var namespace string
+	var oldConfigMap map[string]string
+	var configSpire = map[string]string{
+		"results-from":              "termination-message",
+		"enforce-nonfalsifiability": "spire",
+	}
+
+	c, namespace, oldConfigMap = setupWithFlags(ctx, t, configSpire, requireAllGates(spireFeatureGates))
+	defer resetFlags(ctx, t, c, oldConfigMap)
+
+	knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+	defer tearDown(ctx, t, c, namespace)
+
+	taskRunName := "non-falsifiable-provenance-pass"
+
+	t.Logf("Creating Task and TaskRun in namespace %s", namespace)
+	task := parse.MustParseV1beta1Task(t, fmt.Sprintf(`
+metadata:
+  name: non-falsifiable
+  namespace: %s
+spec:
+  results:
+  - name: foo
+  - name: bar
+  steps:
+  - image: ubuntu
+    script: |
+      #!/usr/bin/env bash
+      sleep 20
```

Comment:
why do we sleep for 20 seconds?

Reply:
This is not needed but does no harm. 
The TR hare has been copied from TestTaskRunModificationSpire where we need the 20sec sleep so that we can modify the TR while the TR is still sleeping and hence fail the task when SPIRE is enabled.